### ISSUE TITLE
Frozen collection follow-ups

### ISSUE THREAD

**Issue by stephentoub** [dotnet/runtime#77891]
https://github.com/dotnet/runtime/pull/77799 adds new FrozenDictionary and FrozenSet types.  These are in decent shape but not yet ready to ship.  Several follow-ups are required:

- [ ] **Construction performance.** The initial use cases for these types was for long-lived processes, where overall it's a better use of resources to spend more time at construction time to optimize the data structures and algorithms used for later data access.  However, the current construction time is one or more orders of magnitude more than for the standard dictionary/set types.  We need to a) invest in driving down those construction costs, and b) consider making the default performance of ToFrozenDictionary/Set as close that of Dictionary/Set as possible via an additional optional parameter to these methods, e.g.
```diff
public static FrozenSet<T> ToFrozenSet<T>(this IEnumerable<T> source, IEqualityComparer<T>? comparer = null,
+    double constructionOptimization = default // pick a better name
);
```
When `0.0`, this would effectively default to creating a FrozenSet/Dictionary that just wrapped a HashSet/Dictionary, providing the same FrozenSet/Dictionary semantics but primarily just as a wrapper for a HashSet/Dictionary, such that the only additional costs were cloning of the data (assuming the Items/Keys/Values properties remain as they are today, they could be lazily-created on first need).  On a scale from 0.0 to 1.0, the larger the double value, the more optimization would be performed at construction time, enabling the caller to choose how much time is spent at construction vs how fast subsequent data access can be made, with 1.0 meaning construction would take the longest in exchange for fastest data access.  Having the default be fast construction would make this the least surprising for casual use and would avoid falling into pits of failure in the average case; consumers willing to spend more time at construction because they know their processes will be long-lived can then opt-in to longer startup times.
- [ ] **Items/Keys/Values properties**.  The current design has Items/Keys/Values properties that return `ImmutableArray<>`.  This is to enable access to the data as a `ReadOnlySpan<>`, indexibility, via refs, and with very fast enumeration.  This does, however, place a requirement on implementations that they either store the keys/values/items separately and in contiguous memory or that implementations maintain a second copy of the data that does so.  If we'd be willing to give up on the `ReadOnlySpan<>` and indexing capabilities, these types should instead be changed to some `FrozenEnumerable<>`-like type that would allow different implementations to vary how the data was stored.
- [ ] **Small collections.** There's currently an optimized implementation for empty collections, but not for other small sizes, e.g. <= 8 elements.  We should investigate adding simple implementations that don't do any hashing or other complicated storage for small sizes.  This includes looking at various code bases / sources of information to determine the most common sizes to be optimized for, e.g. is it worth having a dedicated implementation for collections of just 1 or 2 elements.
- [ ] **Other data structures and algorithms.** Would a perfect hashing implementation yield any benefits?  Would tries be useful in some situations for strings?  Should we consider doing reflection emit code gen in some circumstances?
- [ ] **Source generation.** If/when Roslyn source generators support some kind of mechanism that would allow a source generator to look at constant data in the C# file, it would be beneficial if the optimizations could be performed at compile-time, feeding something into an optional parameter to ToFrozenSet/Dictionary to enable them to be fast, with all the work supplied at build time.  This would likely also require more work on the API to make the existing types extensible, which is something we've actively avoided today.

**Comment by stephentoub** [comment]
cc: @geeknoid

**Comment by geeknoid** [comment]
Perfect hashing ends up being not a great choice. Perfect hashing usually involves a more complex hash function, and you'll end up with lower perf than what we have here. Their main benefit would potentially be in smaller data by not having giant sparse lookup tables.

Another area worth looking into is to further optimize string-based key (the overwhelming favorite kind of key). There are a number of comparers in the current implementation to try and fine-tune how comparison works, but more are possible. For example, comparers that know about 2 character sequences explicitly. The motivation here is to reduce the cost of hashing incoming strings to a minimum.

Perhaps a broader problem that just for the frozen collections, but there is very poor interop between spans and collections. In particular, for string-based dictionaries, it would be terrific to be able to do a lookup using a ReadOnlySpan<char> as a key in addition to a string. This would avoid needing to materialize strings just to do a dictionary lookup. A perfect use for this is a string interning table for incoming REST/gRPC calls. You could go sraight from the incoming buffer, and do a lookup in the dictionary to get the final string object.

**Comment by stephentoub** [comment]
> it would be terrific to be able to do a lookup using a ReadOnlySpan as a key in addition to a string

Yes, this has long been desired, and we currently lack a great answer.  This is tracked by https://github.com/dotnet/runtime/issues/27229.

**Comment by geeknoid** [comment]
Frozen collection state serialization is another interesting possibility.

The idea is to make it possible to serialize the state of a frozen collection to a portable format. A source generator can then generate such serialized blobs to be sucked in by a frozen collection at startup, providing cheap initiatilziation for well-knwon data.

Additionally, said serialized data could actually be passed around in config blobs. So that even state that is config-drivem can have the benefit of low startup time. You serialize the state offline, and send it through the config system. This could in fact provide opportunities for even more aggressive optimizations since startup time would not be an issue.

**Comment by iSazonov** [comment]
> > Frozen collection state serialization is another interesting possibility.
> 
> I intended the "Source generation" part of what I wrote above to cover this. But we'd also need to be _really_ careful here. It'd be fairly easy to accidentally create a vulnerability.

Does this exclude using perfect hash functions at all?

**Comment by stephentoub** [comment]
> > > Frozen collection state serialization is another interesting possibility.
> > 
> > 
> > I intended the "Source generation" part of what I wrote above to cover this. But we'd also need to be _really_ careful here. It'd be fairly easy to accidentally create a vulnerability.
> 
> Does this exclude using perfect hash functions at all?

That statement was unrelated to perfect hashing.

**Comment by iSazonov** [comment]
I mean GetHash using seeding today but perfect hash no. So question can we use it or no.

**Comment by iSazonov** [comment]
> double constructionOptimization = default

It could be an enum like CompressionLevel. It is more user friendly in public API.
Or it could be a dynamic to follow tiered compilation conception - fast startup with follow optimization in background.

**Comment by stephentoub** [comment]
> perfect hash no. So question can we use it or no.

Technically could we? Sure.  Should we?  That would depend on a variety of factors, not the least of which would be demonstrating it was a perf win. And Martin already commented on his experience with their perf in https://github.com/dotnet/runtime/issues/77891#issuecomment-1303687955.

**Comment by iSazonov** [comment]
> That would depend on a variety of factors, not the least of which would be demonstrating it was a perf win. And Martin already commented on his experience with their perf in [#77891 (comment)](https://github.com/dotnet/runtime/issues/77891#issuecomment-1303687955).

I couldn't agree more. A simple example, if we have 256 strings all with different length - perfect hash function can be just the length of the string (and _byte_ lookup table). Is it possible to create something faster?

Complex perfect functions are rather aimed at creating very compact data buffers. If we are not going to store them in the .Net Runtime, we don't need to strive to make them extremely small and hence we can use simpler and faster functions.
We could learn the gperf algorithms, if the license allows it. And create an SG in the first step if that is needed of course.

**Comment by stephentoub** [comment]
> A simple example, if we have 256 strings all with different length

The implementation already does that.

**Comment by omariom** [comment]
Would be very helpful if the dictionary had such an indexer
```C#
public ref readonly KeyValuePair<TKey, TValue> this[int index]

// for the set
public ref readonly T this[int i] 
```
It would allow to store references to entries as integers and later quickly retrieve them.
If returning  KeyValuePair imposes too much on the implementation, then may be separate ```GetValue``` and ```GetKey```?

The idea is to be able to get indexes of the entries and use them to retrieve keys and values.

**Comment by stephentoub** [comment]
> Would be very helpful if the dictionary had such indexer

What's the scenario for that such that the indexer returning a ref to the value is insufficient?

**Comment by omariom** [comment]
You mean this indexer?
```public ref readonly TValue this[TKey key]```

I have a couple of scenarios.

1. Reference data that can be looked up. For example, a list of countries or a list of currencies.
    The referencing data structure would keep entries' indexes, instead of the keys ("UK" or "GBP"). 
    It would also be more efficient in terms of memory and more GC friendly.
    
2.  Another scenario (imaginable in this case) is a networking protocol, that at the start of a session sends the initial snapshot of 
ref data, and later refers to the entries by their indexes instead of the values in the hashset or the keys of the dictionary.

**Comment by geeknoid** [comment]
@omariom Does this make sense in the context of a frozen dictionary? Seems like you're using a dictionary to build up unique name/value pairs, and then when you're done you ditch the name part and access your values by index. Do you have scenarios that routinely require having both key-based and index-based access to the same values on an on-going basis?

**Comment by geeknoid** [comment]
@iSazonov For those cases where the input is known statically, a source generator that has a large suite of tricks to generate custom code seems like a great idea. This is not so much focused on the need for perfect hashing, it could be anything that makes the code fast. 

In the scenarios where I've used frozen collections so far, the set of items in the collections have always come from config or some other dynamic source, so I haven't had the need/desire for a build-time generator. But for parsers which have a well-known set of tokens, there could be value in a highly optimizing source generator. It all depends on the actual gains you could achieve relative to what you already get with the hard-coded frozen collection implementations.

**Comment by geeknoid** [comment]
@stephentoub I have not experimented with purely length-based "hashing", which I think is a fascinating idea.

If there is sufficient variance in the length of the keys in the frozen collection, it should be possible to just use the string length as the hash code, completely eliminating the cost of hashing. It'd be trivial to write a new comparer implementation that just uses string length. I bet that could be a considerable win in many cases.

**Comment by stephentoub** [comment]
>  I have not experimented with purely length-based "hashing", which I think is a fascinating idea.

I spent some time on it, which is where https://github.com/dotnet/runtime/blob/main/src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/LengthBucketsFrozenDictionary.cs came from.

**Comment by geeknoid** [comment]
> > I have not experimented with purely length-based "hashing", which I think is a fascinating idea.
> 
> I spent some time on it, which is where https://github.com/dotnet/runtime/blob/main/src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/LengthBucketsFrozenDictionary.cs came from. There's of course more opportunity to experiment.

Ah, you're one step ahead. I'm surprised this required a new dictionary type, rather than just a string comparer implementation.

**Comment by stephentoub** [comment]
It didn't require it, but the comparer itself added overhead. It might still be useful to have the comparer with the frozen hashtable as a fallback of it didn't meet the criteria for this impl

**Comment by geeknoid** [comment]
Here's another idea:

If an integer set is created that contains a constrained range of values, it would be more efficient to just have an array of bits to hold the set's state.

The idea is to leverage this for sets of enums. FrozenSet<MyEnumType> can then be implemented using simple array lookups rather than hashing.

 I just saw an enum-based set in some production code...

**Comment by AlexRadch** [comment]
I suggested [providing optimized read-only collections for SortedSet and SortedDictionary](#79781). I made a separate topic for discussing the API and other issues related to the implementation of these collections, so as not to interfere with all discussions on different topics in this one topic.

**Comment by Timovzl** [comment]
The [Adaptive Radix Tree](https://db.in.tum.de/~leis/papers/ART.pdf) could be suitable for (ordinal) string keys, especially to implement a sorted set.

A trade-off is retaining the actual string keys vs. keeping only their compacted bytes in the tree.

**Comment by Timovzl** [comment]
> The [Adaptive Radix Tree](https://db.in.tum.de/~leis/papers/ART.pdf)

Come to think of it, the adaptive part is unneccessary for frozen collections. Any radix tree would do.

**Comment by geeknoid** [comment]
@AlexRadch @Timovzl

I think the value in something like SortedSet is that it magically maintains sort order for you as you add/remove items. That magic is not interesting for collections that don't mutate over time. As such, I think the right approach to achieve similar behavior to a sorted collection is to create a normal FrozenSet/Dictionary instance alongside an ImmutableArray with the sorted items/keys (or KeyValuePair). You can then get fast lookup as well as sorted enumeration.

**Comment by AlexRadch** [comment]
@geeknoid yes, I think we can create FrozenOrdered Set/Dictionary with additional Comparer<> and order items/keys by this comparer.  And I think FrozenOrdered Set/Dictionary will support IReadonlyList<> interface also and this[int] indexer.

**Comment by geeknoid** [comment]
@AlexRadch U was actually suggesting that there doesn't need to be a dedicated sorted set/dictionary since it's not really adding any value. The caller can create a frozen set/dictionary with a parallel sorted ImmutableArray on the side. There's no additional efficiency possible by creating a new type in the framework.

**Comment by AlexRadch** [comment]
@geeknoid The actual frozen set/dictionary does not have construction with additional order comparison. We can add ToFrozenSet method with orderComparer parameter,  but they do not implement IReadonlyList<> interface and this[int] indexer. And I think we can not add it to them.

So to support IReadonlyList<> interface and this[int] indexer it should be separated classes.

**Comment by geeknoid** [comment]
Let me state my point again: I don't believe the functionality of SortedFrozenSet and SortedFrozenDictionary should be added to the class framework. These are superfluous data structures. Specifically, if you start with an IEnumerable<T>, application code merely needs to do:

var mySet = myEnumerable.ToFrozenSet();
var myArray = mySet.Items.OrderBy().ToImmutableArray();

And there you go, you can now do O(1) lookups by using mySet, and properly sorted enumeration of the items using myArray. There's little value in packaging these two things together in a single data structure.

**Comment by iSazonov** [comment]
> I don't believe the functionality of SortedFrozenSet and SortedFrozenDictionary should be added to the class framework.

I agree if there is a way to implement compile time optimizations to exclude runtime allocations/construction time lost (important for startup scenarios)/

**Comment by AlexRadch** [comment]
@geeknoid if we add this functionality to the framework we get 2 advantages: creation time and memory allocation. The ordered collections can store keys/values in the right order inside so we will not create a second array.

**Comment by theodorzoulias** [comment]
> So to support `IReadonlyList<>` interface and `this[int]` indexer it should be separated classes.

@AlexRadch isn't this problematic for a `FrozenDictionary<TKey, TValue>` with the `Tkey` being of type `int`? I would expect compile-time errors because of the ambiguity between

```C#
TValue IReadOnlyDictionary<TKey, TValue>.this[TKey key]
```

and

```C#
KeyValuePair<TKey, TValue> IReadOnlyList<KeyValuePair<TKey, TValue>>.this[int index]
```

Maybe the derived classes could implement the `IReadOnlyList<>` interface, so casting from `FrozenDictionary<>` to `IReadOnlyList<>` would succeed, and the `this[int]` indexer would be available through the interface for `TKey` different than `int`. But what is the scenario that the `this[int]` indexer would be useful? Isn't ordered enumeration enough?

**Comment by AlexRadch** [comment]
> @AlexRadch isn't this problematic for a `FrozenDictionary<TKey, TValue>` with the `TKey` being of type `int`? I would expect compile-time errors because of the ambiguity between
> Maybe the derived classes could implement the `IReadOnlyList<KeyValuePair<TKey, TValue>>` interface, so casting from `FrozenDictionary<>` to `IReadOnlyList<>` would succeed, and the `this[int]` indexer would be available through the interface. But what is the scenario that the `this[int]` indexer would be useful? Isn't ordered enumeration enough?

Yes, you are right, so I think the `IList<>` interface will be enough. I think Frozen OrderedDictionary should have `public ref readonly TValue GetValueRefByIndex(int index)` method to get the reference to a value by index.

**Comment by theodorzoulias** [comment]
@AlexRadch personally I've never used an [`OrderedDictionary`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.specialized.ordereddictionary). I don't think that it's a useful data structure. I guess it can be useful in case you want to paginate a dictionary, and fetch efficiently the items 9,901-10,000. That's the only usage scenario that I can think of.

**Comment by danmoseley** [comment]
Is the scenario of frozen keys, mutable values supported by the new collections? Eg., in this scenario (on Kestrel hot path) a dictionary lookup is basically hand generated (these days maybe Roslyn would generate similar code from a switch statement) and the value may change. https://github.com/danmoseley/aspnetcore/blob/41627fe50a7a71e395f2fdfccce9f8cbff10a49a/src/Shared/HttpSys/RequestProcessing/RequestHeaders.Generated.cs#L968

**Comment by stephentoub** [comment]
> Is the scenario of frozen keys, mutable values supported by the new collections?

You can't mutate what value the dictionary/set stores for a key.  If that value is a reference type, though, or if it contains a reference type, you can of course mutate the referenced instance.  In other words, the immutability here is shallow.

**Comment by danmoseley** [comment]
Yeah in this case it's a struct. OK thanks.

**Comment by Timovzl** [comment]
@danmoseley I have definitely seen someone use GetValueRefOrNullRef() and mutate a struct value in a frozen dictionary somewhere…

**Comment by stephentoub** [comment]
>  I have definitely seen someone use GetValueRefOrNullRef() and mutate a struct value in a frozen dictionary somewhere…

That returns a ref readonly. Using a ref readonly to erroneously mutate the collection requires unsafe code. With unsafe code you can violate almost any kind of constraint, including immutability.

**Comment by Timovzl** [comment]
> Using a ref readonly to erroneously mutate the collection requires unsafe code.

Ahhh, so that was the trick. Indeed, they used a method from the `Unsafe` class to do so, I now recall. So that was definitely not by design.

**Comment by mjsabby** [comment]
Perfect hashing doesn't have to be complex -- take a look at https://gist.github.com/mjsabby/d7baf64d524176ee35f69ffa2c278179

Implements Hash Displace algorithm. I would implement Frozen Sets and Dictionaries with perfect hashing. In fact, for a bunch of my projects they are serialized as FlatBuffers.

```csharp
        [MethodImpl(MethodImplOptions.AggressiveOptimization)]
        public static int Lookup(ReadOnlySpan<byte> key, ReadOnlySpan<int> values, ReadOnlySpan<int> seeds)
        {
            ulong hash;

            unsafe
            {
                fixed (byte* ptr = &MemoryMarshal.GetReference(key))
                {
                    hash = Hash64(ptr, key.Length);
                }
            }

            var size = (ulong)values.Length;
            var index = hash & (size - 1);
            var seed = seeds[(int)index];

            if (seed < 0)
            {
                return values[0 - seed - 1];
            }

            var x = (ulong)seed + hash;
            x ^= x >> 12;
            x ^= x << 25;
            x ^= x >> 27;

            index = (x * 0x2545F4914F6CDD1D) & (size - 1);
            return values[(int)index];
        }
```

**Comment by YoniFeng** [comment]
> Perfect hashing ends up being not a great choice. Perfect hashing usually involves a more complex hash function, and you'll end up with lower perf than what we have here. Their main benefit would potentially be in smaller data by not having giant sparse lookup tables.

I haven't taken the time to look at the Frozen collection's code (would love TL;DR/gist, but I'll have to take a good look regardless), but a PHF(perfect hash function) definitely has value.
I'm surprised you seem to dismiss it for poor key lookup performance (did I infer wrong?), I'd think the only reason to avoid a PHF during runtime is the construction time can be very slow.

> For those cases where the input is known statically, a source generator that has a large suite of tricks to generate custom code seems like a great idea. This is not so much focused on the need for perfect hashing, it could be anything that makes the code fast.

A source generator for static sets could be fantastic. It can make top-of-the-line lookup perf easily accessible.
It would essentially be like an integrated [gperf](https://www.gnu.org/software/gperf/). An ~example is [rust-phf](https://github.com/rust-phf/rust-phf) (except it doesn't allow customizing the underlying hash function and comparer, is _minimal_, and Rust's default hasher is unnecessarily slow because it provides cryptographic guarantees that aren't needed for static sets).

About _minimal_: I'm yet to do thorough benchmarking of the PHF subject (planning to), but my hunch is having a minimal PHF results in _worse_ performance.
It optimizes for space efficiency (binary size / table stored in data segment) over time efficiency (lookup speed), because a non-keyword hitting an empty slot can be rejected faster (in the case of strings - faster or no string comparison).

> Is the scenario of frozen keys, mutable values supported by the new collections? Eg., in this scenario (on Kestrel hot path) a dictionary lookup is basically hand generated (these days maybe Roslyn would generate similar code from a switch statement) and the value may change. https://github.com/dotnet/aspnetcore/blob/41627fe50a7a71e395f2fdfccce9f8cbff10a49a/src/Shared/HttpSys/RequestProcessing/RequestHeaders.Generated.cs#L968
> 
> Edit: never mind, it uses fields for the values. But I'm interested in the answer anyway.

@danmoseley
Can you give a small elaboration on HTTP request keyword/verb matching is used in Kestrel? Why does the value need to be mutated? (Like above - I need to find the time to dig in to the code myself..).
I was looking for this the other day, because I wanted to see how Kestrel/ASP.NET match HTTP keywords and/or common request headers.

This the sort of use case I'd expect could see great benefit from a source-generated PHF set/dictionary.
Assessing whether a string is a reserved keyword (e.g. in compilers) is exactly the use-case where a compile-time PHF-based structure can shine.

Come to think of it - how does Roslyn do keyword matching?

**Comment by danmoseley** [comment]
@YoniFeng  I’m not familiar with that code. Of course any contributions or analysis you may have are welcome.

**Comment by rampaa** [comment]
> Source generation
If/when Roslyn source generators support some kind of mechanism that would allow a source generator to look at constant data in the C# file and augment ToFrozenSet/Dictionary call sites in some manner, it would be beneficial if the optimizations could be performed at compile-time, feeding something into an optional parameter to ToFrozenSet/Dictionary to enable them to be fast, with all the work supplied at build time. This would likely also require more work on the API to make the existing types extensible, which is something we've actively avoided today.

Is this on the radar for .NET 8?

**Comment by stephentoub** [comment]
> Is this on the radar for .NET 8?

No.  It requires exposing the details of the abstraction, and we want some bake time on that before we do.  It's something we'll consider for .NET 9.

**Comment by stephentoub** [comment]
Everything for 8 is done

**Comment by andrewjsaid** [comment]
Another way to deliver on the premise of improving lookup time at a potential one-time cost of construction would be some kind of mechanism which tracks usage patterns. This could be provided either at construction time (e.g. an array of key lookups) or by enabling some kind of sampling to be run to track lookup keys.

The frozen collection would then have a method to re-optimize to speed up lookups based on the usage pattern.

Here are some metrics which would be great to know ahead of time **for stringly keyed FrozenDictionaries** in order to create more optimal and targeted lookups:
1. Are most lookups for the same key or small subset of keys? - e.g. a `FrozenDictionary` where the key is country code and most of the time it's the same country code being looked up. Maybe there could be two dictionaries - a faster one with fewer items and a regular one with the entire set.
2. Changing the number of buckets to ensure that frequently accessed keys are not in the same bucket.
3. Rotating the order of keys within buckets to sort by most frequently accessed
4. Not using max length checks if we know them to be ineffective with that particular data set.
5. Are we ignoring case unnecessarily most of the time? If so maybe there could be two dictionaries here too.

I'm sure that some of these ideas aren't suitable or are slow to compute, but equally I'm sure there are others I haven't thought of which could be perfect optimization.

I guess the problem with this suggestion is that I'm not entirely sure how `FrozenDictionaries` are used in the wild, so maybe it's better to have this conversation once some common scenarios exist.

**Comment by andrewjsaid** [comment]
> Would tries be useful in some situations for strings? Should we consider doing reflection emit code gen in some circumstances?

After I saw that routing in ASP.NET Core uses reflection emitted Trie, I also spent some time attempting to implement an alternative to stringly-keyed ASCII `FrozenDictionary` in a similar way. After lots of attempts, I no longer think it's possible for Tree structures to be faster than Hash based data structures in general purpose use-cases. However the IL-Emit Trie did indeed come out faster than the `FrozenDictionary` in my small experiments.

https://github.com/andrewjsaid/trie

**Comment by rampaa** [comment]
>  It's something we'll consider for .NET 9.

Is it being considered for .NET 9?

**Comment by stephentoub** [comment]
> > It's something we'll consider for .NET 9.
> 
> Is it being considered for .NET 9?

Unlikely at this point.

**Comment by sanghel-orbyta** [comment]
Any chance of considering adding something like:
- `FrozenSortedSet<T>`
- `FrozenSortedDictionary<T>`
- `IEnumerable<T>.ToFrozenSortedSet<T>`
- `IEnumerable<T>.ToFrozenSortedDictionary<T>`
?

Perhaps via analogous methods such as those for `.ToImmutableSortedDictionary/Set()`

**Comment by stephentoub** [comment]
> Any chance of considering adding something like:
> 
> * `FrozenSortedSet<T>`
> * `FrozenSortedDictionary<T>`
> * `IEnumerable<T>.ToFrozenSortedSet<T>`
> * `IEnumerable<T>.ToFrozenSortedDictionary<T>`
>   ?
> 
> Perhaps via analogous methods such as those for `.ToImmutableSortedDictionary/Set()`

What functionality are you hoping for here? Specifically just that the exposed Items and Key/Value arrays are sorted and that enumerating happens in that order as well?

**Comment by sanghel-orbyta** [comment]
@stephentoub exactly.

I have a ConcurrentDictionary that I accumulate into and then need to read from ordered by key, as quickly as possibile (using this for some UDP traffic re-ordering, the dictionary is a `ConcurrentDictionary<long, (int packetOffset, int packetLenght)>`, that keeps track of offsets in a `Memory<byte>` buffer... linear, I know, but I'm not expert enough to know how to do this with a ring buffer with pooled memory alocations avoiding GC...).

Right now I'm using `.ToImmutableSortedDictionary()`, but from what I've understood, a FrozenDictionary would be quicker to enumerate.

**Comment by stephentoub** [comment]
>> What functionality are you hoping for here? Specifically just that the exposed Items and Key/Value arrays are sorted and that enumerating happens in that order as well?

> @stephentoub exactly.

In that case, I don't think this warrants or necessitates new types. We could just have the creation of a FrozenDictionary/Set sort the arrays at construction, either using a type's built-in notion of ordering (via IComparable) which could be done without any API additions, or via an extra overload that accepts a Comparer to use for that purpose.

**Comment by sanghel-orbyta** [comment]
> In that case, I don't think this warrants or necessitates new types. We could just have the creation of a FrozenDictionary/Set sort the arrays at construction

That would definitely work, thanks for the feedback @stephentoub .

**Comment by molinch** [comment]
Using Benchmark.NET I compared `FrozenSet<int>` vs `HashSet<int>` with small sets (5, 10, 20, 30).
I do realize that it's a niche scenario but I was still not expecting these results.

In my benchmark, FrozenSet shines from around 20 items.
I was under the impression that it would always outperform a regular HashSet since it can choose different internal implementations at construction time. 
Is that by design?

<img width="834" height="341" alt="Image" src="https://github.com/user-attachments/assets/173ce3dc-e6f5-47a3-92c8-d410d9b9e35d" />

**Comment by stephentoub** [comment]
> In my benchmark

Can you share it?

**Comment by molinch** [comment]
> > In my benchmark
> 
> Can you share it?

Sure!

```csharp
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.Collections.Frozen;

BenchmarkRunner.Run<LookupBenchmarks>();

public class LookupBenchmarks
{
    [Params(10, 20, 30)]
    public int CollectionSize { get; set; }

    private HashSet<int> _hashSet;
    private FrozenSet<int> _frozenSet;

    [GlobalSetup]
    public void SetUp()
    {
        _hashSet = Enumerable.Range(0, CollectionSize).ToHashSet();
        _frozenSet = Enumerable.Range(0, CollectionSize).ToFrozenSet();
    }

    [Benchmark]
    public void LookupHashSet()
    {
        for (var i = 0; i < CollectionSize; i++)
            _ = _hashSet.Contains(i);
    }

    [Benchmark]
    public void LookupFrozenSet()
    {
        for (var i = 0; i < CollectionSize; i++)
            _ = _frozenSet.Contains(i);
    }
}
```

**Comment by molinch** [comment]
@stephentoub Do you need anything else from me? I know it’s pretty niche, but it’d be awesome if FrozenSet could be a jack of all trades - and a good one :)

**Comment by rampaa** [comment]
> > > It's something we'll consider for .NET 9.
> > 
> > Is it being considered for .NET 9?
> 
> Unlikely at this point.

Is it being considered for .NET 11? 🥲

**Comment by stephentoub** [comment]
> [@stephentoub](https://github.com/stephentoub) Do you need anything else from me? I know it’s pretty niche, but it’d be awesome if FrozenSet could be a jack of all trades - and a good one :)

I haven't been able to look yet. Sorry.
